### PR TITLE
First draft of workbench

### PR DESCRIPTION
**Pull request by raphaelmue**
See issue #67

### CODE REVIEW

**Comment by raphaelmue** [comment]
ToDo:

- [ ] Update code of published widgets
- [ ] Delete published widget
- [ ] Checks on code
- [ ] Add draft and published to dashup
- [ ] Modify category and tags

File: de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+                    model.addAttribute("current", widget);
+                    DashupService.getInstance().getTagsByWidget(widget);
+                    model.addAttribute("currentTags", widget.getTags());
+                    break;
+                }
+            }
+        });
+    }
+
+    @RequestMapping(value = "/draft/{draftId}/deleteDraft")
```
Comment (fidsusj) [inline de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java]: We should maybe make sure, that the user making the delete request is the user, who created the panel. We do not want to have Jannik delete all our widgets mid presentation.

File: de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+                                     @PathVariable(value = "publishedId") int publishedId) throws SQLException {
+        User user = LocalStorage.getInstance().getUser(request, token);
+        if (user != null) {
+            DashupService.getInstance().deleteDraft(publishedId);
+            return "redirect:/workbench/#deletedWidget";
+        }
+        return "redirect:/login";
+    }
+
+    @RequestMapping(value = "/draft/{draftId}/changeInformation", method = RequestMethod.POST)
```
Comment (fidsusj) [inline de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java]: Add a condition, whether the user making the request is the owner of the widget

File: de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        User user = LocalStorage.getInstance().getUser(request, token);
+        if (user != null) {
+            Draft draft = new Draft();
+            updateInformation(draft, draftId, name, shortDescription, description, category, tagsJSON);
+            return "redirect:/workbench/draft/" + draftId + "#changedInformation";
+        }
+        return "redirect:/login";
+    }
+
+    @RequestMapping(value = "/published/{publishedId}/changeInformation", method = RequestMethod.POST)
```
Comment (fidsusj) [inline de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java]: same goes for this

File: de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        JSONArray json = new JSONArray(URLDecoder.decode(tagsJSON, StandardCharsets.UTF_8.name()));
+        for (int i = 0; i < json.length(); i++) {
+            widget.getTags().add(new Tag(json.getJSONObject(i).getInt("id"), json.getJSONObject(i).getString("name")));
+        }
+
+        DashupService.getInstance().updateWidgetInformation(widget);
+    }
+
+    @RequestMapping(value = "/draft/{draftId}/changeCode", method = RequestMethod.POST)
+    public String handleChangeDraftCode(@CookieValue(name = "token", required = false) String token,
```
Comment (fidsusj) [inline de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java]: or here ^^

File: de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        if (user != null) {
+            Draft draft = new Draft();
+            updateCode(draftId, codeSmall, codeMedium, codeLarge, draft);
+            return "redirect:/workbench/draft/" + draftId + "#changedCode";
+        }
+        return "redirect:/login";
+    }
+
+    @RequestMapping(value = "/published/{publishedId}/changeCode", method = RequestMethod.POST)
+    public String handleChangeWidgetCode(@CookieValue(name = "token", required = false) String token,
```
Comment (fidsusj) [inline de.dashup.application/src/main/java/de/dashup/application/controllers/WorkbenchController.java]: You know what to do. Goes for the rest of the controller handlers as well. I will not comment this issue further.

File: de.dashup.application/src/main/resources/i18n_de.properties [hunk 6]
```diff
@@ -66,13 +66,49 @@ i18n.save=Speichern
...
 i18n.successChangedEmail=E-Mail Adresse erfolgreich ge�ndert!
 i18n.generalError=Etwas ist schief gelaufen. Bitte versuchen Sie es sp�ter noch einmal.
 i18n.errorInvalidEmail=Die eingegebene E-mail Adresse ist invalide!
 i18n.changeUsername=Benutzernamen �ndern
 i18n.errorUserNameIsAlreadyRegistered=Dieser Benutzername ist bereits leider vergeben!
 i18n.company=Firma
 i18n.bio=Bio
+i18n.successChangedInformation=Informationen erfolgreich ge�ndert!
+i18n.warningDraftNotValid=Der Entwurf ist ung�ltig. Es fehlen einige Informationen.
+i18n.confirmPublishDraft=Sind Sie sicher, dass Sie diesen Entwurf ver�ffentlichen wollen? Sie k�nnen diese Aktion nicht r�ckg�ngig machen und nach der Ver�ffentlichung k�nnen Sie Ihr Widget nicht mehr �ndern.
```
Comment (fidsusj) [inline de.dashup.application/src/main/resources/i18n_de.properties]: But I can delete publications right?

File: de.dashup.model/src/main/java/de/dashup/model/service/Validator.java [hunk 7]
```diff
@@ -1,24 +1,70 @@
...
+
+    static {
+        whitelist = new Whitelist();
+        whitelist.addAttributes("dashup-button", "text", "mode", "api", "params", "consumers", "producers", "disabled");
+        whitelist.addAttributes("dashup-chart", "title", "category", "data");
+        whitelist.addAttributes("dashup-checkbox", "value", "checked", "disabled");
+        whitelist.addAttributes("dashup-component", "name");
+        whitelist.addAttributes("dashup-display", "label", "displayedText", "quantity");
+        whitelist.addAttributes("dashup-grid-layout", "name");
+        whitelist.addAttributes("dashup-list", "selectable");
```
Comment (fidsusj) [inline de.dashup.model/src/main/java/de/dashup/model/service/Validator.java]: Is the name attribute as well allowed for the subclasses of component?

File: de.dashup.shared/src/main/java/de/dashup/shared/Widget.java [hunk 8]
```diff
@@ -2,12 +2,15 @@
...
+import java.util.HashSet;
+import java.util.Set;
+
+public class Widget extends DatabaseWidget implements Comparable<Widget> {
 
     public enum Size {
         SMALL("small", "m2 s6"),
         MEDIUM("medium", "m4 s12"),
-        LARGE("large", "m5 s12");
+        LARGE("large", "m6 s12");
```
Comment (fidsusj) [inline de.dashup.shared/src/main/java/de/dashup/shared/Widget.java]: I once changed it to m5 to have some space between the widgets in the central dashboard.

**Comment by fidsusj** [comment]
And in the metadata section the placeholder for the tags is out of the display range of the input field.

**Comment by raphaelmue** [comment]
> * [x]  The page is unfortunately not displayed correctly on my laptop. The menu that should be on the side is horizontal at the top of the page. For the actual content you have to scroll relatively far down. Therefore it would be good to have a closer look at the layout again.

This issue is fixed by clearing cache and reloading CSS files.
 
> * [x]  Saving general information results in an error (SQLSyntaxErrorException), because the column "category" is not known (according to error message)

Did you use the new database scheme? It sounds like, you did not update the panel table.

File: de.dashup.shared/src/main/java/de/dashup/shared/Widget.java (hunk 8 above)
Comment (raphaelmue) [reply de.dashup.shared/src/main/java/de/dashup/shared/Widget.java]: I know, and I fixed this issue 😉

**Comment by raphaelmue** [comment]
> * if I change the widget size, I do not want my coding to disappear suddenly (users will try to find out, which size suits best for their widget by switching during development)

Users need to implement all 3 sizes. You cannot publish a widget without having all 3 sizes implemented. Why should all sizes code be equal when developing, but not when publishing? I think developers are able to use copy and paste. 

> * make widget size constant s6 m6 l6 xl6 in workbench

Will be done

> * the user_draft table is not deleted so far, but I guess it will?

It is already deleted on remote database, but I forgot to delete in the SQL file. 

> * can you make the code input field a default height? (If not, then add some extra newlines)

Will be done

> * move the + button to the widget list on the right

CSS does not want it, and I cannot do anything about it. 

> * make tab usable when trying to tab in code (current behavior: tabs to next input control like save button)

Will be done, if possible

> * please adopt the styling of the cards in the central dashboard (especially remove the huge padding)

Will be done.

> * give an error message when a widget could not be saved and leave the code as it is in the input field ()   (users will rage if they have to type in the whole code again if they did one single mistake)

Code will be saved even thought there are errors, but you cannot publish if code is not valid. If something goes wrong in backend, you can't save it anyways.

> * java.sql.SQLSyntaxErrorException: Unknown column 'category' in 'field list' when trying to save metadata

Forgot to add this is the SQL file. 

> * I guess every kind of input is allowed for the tags? (Even some bu*****t like aodaod)?

I actually prevented to enter invalid tags. So you cannot save these.

> * When having added a widget (not published yet) to the dashboard and deleting it (regardless if published or not), an error occurs (Cannot delete or update a parent row: a foreign key constraint fails (`dashup_dev`.`sections_panels`, CONSTRAINT `section_panels_panel_id` FOREIGN KEY (`panel_id`) REFERENCES `panels` (`id`)))

I don't know, why this is not working for you, I can delete this widget from my dashup. Maybe your database scheme is not correct.

>   Expected behavior: Panel is deleted both in workbench and dashboard

Will be done. 

> * When adding widget multiple times, no error message occurs, that widget is already added to the dashboard (regardless whether section is the same or not)

Why is this an error?

> * When updating the category or tags of a draft, an error occurs (Unknown column 'category' in 'field list')

I will send an update with a correct SQL file.

**Comment by fidsusj** [comment]
Maybe you could add the + button on the left side instead? 

Why would a user add the same widget multiple times? But well, we could leave it like that. The time investment would not be worth it.

**Comment by raphaelmue** [comment]
That's the correct db scheme
[dashup_dev.zip](https://github.com/raphaelmue/dashup/files/3201226/dashup_dev.zip)

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java [hunk 9]
```diff
@@ -392,4 +393,206 @@ public void processLayoutModeChanges(LayoutModeStructureDTO layoutModeStructureD
...
+        if (widget.getDescription() != null) {
+            values.put("descriptions", widget.getDescription());
+        }
+        if (widget.getCategoryObject() != null) {
+            values.put("category", widget.getCategory());
+        }
+
+        this.database.update(Database.Table.PANELS, whereParameters, values);
+
+        if (widget.getTags().size() > 0) {
```
Comment (joshuaschu) [inline de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: You are using the widget tag count to test if an update is necessary, but if i delete all tags this count is 0 and you are updating nothing even if there are some tags that should be deleted.

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java [hunk 10]
```diff
@@ -392,4 +393,206 @@ public void processLayoutModeChanges(LayoutModeStructureDTO layoutModeStructureD
...
+            this.updateWidgetTags(widget, widget.getTags());
```
Comment (joshuaschu) [inline de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: Somehow it is not possible to delete the tags. As discussed...

File: de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java [hunk 11]
```diff
@@ -392,4 +393,206 @@ public void processLayoutModeChanges(LayoutModeStructureDTO layoutModeStructureD
...
+    }
+
+    private void checkWidget(Widget widget) throws MissingInformationException, InvalidCodeException {
+        if ((Validator.isNullOrEmpty(widget.getName()) ||
+                Validator.isNullOrEmpty(widget.getDescription()) ||
+                Validator.isNullOrEmpty(widget.getShortDescription()) ||
+                Validator.isNullOrEmpty(widget.getCode(Widget.Size.SMALL)) ||
+                Validator.isNullOrEmpty(widget.getCode(Widget.Size.MEDIUM)) ||
+                Validator.isNullOrEmpty(widget.getCode(Widget.Size.LARGE)))) {
+            throw new MissingInformationException(Draft.class);
```
Comment (joshuaschu) [inline de.dashup.model/src/main/java/de/dashup/model/service/DashupService.java]: If there is a missing info... maybe you can pass the spot where the info is missing via this exception and mark the field in the frontend or something? Lets discuss that later in our team meeting...

**Pull request merged**